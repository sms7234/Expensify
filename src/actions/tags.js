import database from '../firebase/firebase';

export const setTags = (tags) => ({
  type: "SET_TAGS",
  tags
});

export const startSetTags = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags`)
      .once('value')
      .then((snapshot) => {
        const tags = [];
        snapshot.forEach((child) => {
          tags.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setTags(tags));
      });
  };
};

export const addTag = (tag) =>({
  type: "ADD_TAG",
  tag
});

export const startAddTag = (tagData={}) => {
  return (dispatch, getState) => {
    const {
      tag='',
      description=''
    } = tagData;
    const tagItem={tag,description};
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/tags`).push(tagItem).then((ref)=>{
      dispatch(addTag({
        id: ref.key,
        ...tagItem
      }));
    });
  };
};

export const editTag = (id, updates) => ({
  type: 'EDIT_TAG',
  id,
  updates
});

export const startEditTag = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).update(updates).then(()=>{
      dispatch(editTag(id,updates));
    });
  };
};

export const removeTag = (id) => ({
  type: 'REMOVE_TAG',
  id
});

export const startRemoveTag = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/tags/${id}`).remove().then(()=>{
      dispatch(removeTag(id));
    });
  };
};
