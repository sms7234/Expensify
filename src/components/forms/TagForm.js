import React from 'react';

export default class TagForm extends React.Component{
  constructor(props){
    super(props);
    this.state={
      tag: props.tags ? props.tags.tag: '',
      description: props.tags ? props.tags.description:'',
      error:''
    };
  }
  onTagChange = (e) => {
    const tag = e.target.value;
    if(tag.length > 0){
      this.setState(()=> ({tag}));
    };
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(()=>({description}))
  };
  onSubmit = (e) => {
    e.preventDefault();
    if(!this.state.tag){
      this.setState(()=> ({error: 'Please fill in tag'}));
    } else {
      this.setState(()=>({error:''}));
      this.props.onSubmit({
        tag: this.state.tag,
        description: this.state.description
      })
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error!=='' && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="tag name"
          className="text-input"
          value={this.state.tag}
          onChange={this.onTagChange}
          autoFocus
        />
        <textarea
          placeholder="Add description for your tag"
          className="textarea"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <button className="button">
          Save
        </button>
      </form>
    )
  }
}
