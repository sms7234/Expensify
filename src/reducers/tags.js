export default (state=[], action) => {
  switch(action.type){
    case 'SET_TAGS':
      return action.tags;
    case 'ADD_TAG':
      return state.concat(action.tag);
    case 'EDIT_TAG':
      return state.map((item) => {
        if (item.id === action.id){
          return {
            ...item,
            ...action.updates
          }
        } else {return item}
      });
      case 'REMOVE_TAG':
        return state.filter(({id}) => id!==action.id);
    default:
      return state;
  }
}
