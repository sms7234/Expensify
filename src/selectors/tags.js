export default (tags, {tag}) => {
  return tags.filter((item) => {
    const textMatch = item.tag.toLowerCase().includes(tag.toLowerCase());
    return textMatch;
  }).sort((a,b)=> {
    if(a.tag < b.tag){
      return -1;
    } else{
      return 1;
    }
  });
};
