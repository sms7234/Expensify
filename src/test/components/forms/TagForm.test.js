import React from 'react';
import {shallow} from 'enzyme';
import TagForm from '../../../components/forms/TagForm';
import tags from '../../fixtures/tags';

test('should render Tag Form with no data', () => {
  const wrapper = shallow(<TagForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render TagForm with data', () => {
  const wrapper = shallow(<TagForm tag={tags[0]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid submission', () => {
  const wrapper = shallow(<TagForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set tag value on change', () => {
  const value = "new tag";
  const wrapper= shallow(<TagForm />);
  wrapper.find('input').simulate('change', {
    target: {value}
  });
  expect(wrapper.state('tag')).toBe(value);
});

test('should set tag value on change', () => {
  const value = "new description";
  const wrapper= shallow(<TagForm />);
  wrapper.find('textarea').simulate('change', {
    target: {value}
  });
  expect(wrapper.state('description')).toBe(value);
});
