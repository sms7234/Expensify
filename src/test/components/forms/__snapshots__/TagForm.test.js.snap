// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Tag Form with no data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="tag name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your tag"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render TagForm with data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="tag name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your tag"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render error for invalid submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="tag name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your tag"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;

exports[`should render error for invalid submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please fill in tag
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="tag name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add description for your tag"
    value=""
  />
  <button
    className="button"
  >
    Save
  </button>
</form>
`;
